### PR TITLE
*: fix bug when unsigned pk meets feedback ranges

### LINKED ISSUE #10242: invalid range,range.start should be smaller than range.end
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do?
Execute query output errors .
SQL:
select count(*) from transaction_stats_tab WHERE (whs_id = 'IDG' AND transaction_type = 2
2. What did you expect to see?
Execute successful.

3. What did you see instead?
ERROR 1105(HY0000):other error: unknown error StringError("[src/coprocessor/dag/executor/table_scan.rs:56]: invalid range,range.start should be smaller than range.end, but got [[116, 128, 0, 0, 0, 0, 0, 19, 96, 95, 114, 254, 66, 203, 60, 110, 208, 160, 57],[116, 128, 0, 0, 0, 0, 0, 19, 96, 95, 114, 0, 28, 248, 3, 250, 72, 160, 57])")

4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?
Release Version: v2.1.0-5-g4dad722
Git Commit Hash: 4dad722bbf573050477328ce8ff5bcb7f1a1c7f1
Git Branch: release-2.1
UTC Build Time: 2018-12-07 02:29:19
GoVersion: go version go1.11.2 linux/amd64
Race Enabled: false
TiKV Min Version: 2.1.0-alpha.1-ff3dd160846b7d1aed9079c389fc188f7f5ea13e
Check Table Before Drop: false

### PR DESCRIPTION
**Pull request by winoros**
### What problem does this PR solve? 

fix #10242

### What is changed and how it works?

Port some codes from #7921 

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change

Related changes

 - Need to be included in the release note

### CODE REVIEW

File: statistics/histogram.go [hunk 1]
```diff
@@ -508,7 +508,15 @@ func validRange(ran *ranger.Range) bool {
 	if ran.LowVal[0].Kind() == types.KindBytes {
 		low, high = ran.LowVal[0].GetBytes(), ran.HighVal[0].GetBytes()
 	} else {
-		low, high = codec.EncodeInt(nil, ran.LowVal[0].GetInt64()), codec.EncodeInt(nil, ran.HighVal[0].GetInt64())
+		var err error
+		low, err = codec.EncodeKey(nil, nil, ran.LowVal[0])
```
Comment (winoros) [inline statistics/histogram.go]: In release 2.1, only int primary key will go to here. So pass a nil `stmtCtx` is okay.

File: distsql/request_builder.go [hunk 2]
```diff
@@ -183,7 +183,6 @@ func TableRangesToKVRanges(tid int64, ranges []*ranger.Range, fb *statistics.Que
 	if fb == nil || fb.Hist() == nil {
 		return tableRangesToKVRangesWithoutSplit(tid, ranges)
 	}
-	ranges = fb.Hist().SplitRange(ranges)
```
Comment (lamxTyler) [inline distsql/request_builder.go]: Split it in `table_reader`?

File: statistics/histogram.go (hunk 1 above)
Comment (lamxTyler) [reply statistics/histogram.go]: Seems not, in feedback.go:1106, it could be other types.

File: distsql/request_builder.go (hunk 2 above)
Comment (winoros) [reply distsql/request_builder.go]: Removed by mistake when remove debugging log.

File: statistics/histogram.go (hunk 1 above)
Comment (winoros) [reply statistics/histogram.go]: No, `validRange` is only called by `SplitRange`.
And `SplitRange` is only call when converting table/index range to kv range or `dumpRangeFeedback` so other type won't be passed as a param.

**Comment by shenli** [comment]
Could you add some test cases?

File: statistics/histogram.go (hunk 1 above)
Comment (lamxTyler) [reply statistics/histogram.go]: Yes, but in `dumpRangeFeedback` it could be other types.

File: distsql/request_builder.go (hunk 2 above)
Comment (eurekaka) [reply distsql/request_builder.go]: Why should we move this `SplitRange` to `TableReaderExecutor::Open`?

**Comment by winoros** [comment]
@eurekaka 
We need to move it because of the following reasons.
1. In `TableRangesToKVRanges`, we encode the `ranger.Range` to `kv.KeyRange`. Since we store unsigned pk as int pk in physical kv layer in fact, unsigned ones are stored before signed ones in fact.
1. But the `SplitRange` will put the unsigned ones after signed ones since it stores logical value.
1. And if the reader doesn't set `KeepOrder` true. We do give `TableRangesToKVRanges` a full range contain both unsigned part and signed part. Then we got a wrong `Range` slice after `Histogram.SplitRange`.

**Comment by winoros** [comment]
When adding tests, finding that though the response can be sent correctly, but the feedback cannot be updated correctly. Need more changes to correct it.

File: executor/table_reader.go [hunk 3]
```diff
@@ -96,6 +96,9 @@ func (e *TableReaderExecutor) Open(ctx context.Context) error {
 	}
 
 	e.resultHandler = &tableResultHandler{}
+	if e.feedback != nil && e.feedback.Hist() != nil {
+		e.ranges = e.feedback.Hist().SplitRange(e.ranges)
+	}
```
Comment (eurekaka) [inline executor/table_reader.go]: Got it, please add the comment here to inform that `SplitRange` must be placed before `splitRanges` and explain the reasons.

File: statistics/histogram.go (hunk 1 above)
Comment (zz-jason) [reply statistics/histogram.go]: By passing a `StmtCtx` to `SplitRange()`, we can safely handle all the situations.

**Comment by winoros** [comment]
@lamxTyler @eurekaka Addressed, PTAL

**Comment by eurekaka** [comment]
@winoros CI failed, PTAL

**Comment by winoros** [comment]
@eurekaka fixed

File: statistics/histogram.go (hunk 1 above)
Comment (lamxTyler) [reply statistics/histogram.go]: PTAL @winoros

**Comment by winoros** [comment]
@lamxTyler Fixed

File: statistics/feedback.go [hunk 4]
```diff
@@ -1164,7 +1164,7 @@ func (q *QueryFeedback) dumpRangeFeedback(h *Handle, ran *ranger.Range, rangeCou
 			ran.HighVal[0] = getMaxValue(q.hist.tp)
 		}
 	}
-	ranges := q.hist.SplitRange([]*ranger.Range{ran})
+	ranges := q.hist.SplitRange(nil, []*ranger.Range{ran})
```
Comment (lamxTyler) [inline statistics/feedback.go]: Should we use the `sc` in line 1145?

Comment (winoros) [reply statistics/feedback.go]: index value in histogram are stores as bytes. So use `nil` is okay?

Comment (lamxTyler) [reply statistics/feedback.go]: It may also be column, and we support column types like `timestamp`, so `nil` may not suit here.

Comment (winoros) [reply statistics/feedback.go]: @lamxTyler fixed

**Pull request merged**